File: LeetCode root folder/Append K Integers With Minimal Sum.cpp
//greedy solution:111
//ms
//Beats
//56.90%

class Solution {
	public:
			long long minimalKSum(vector<int>& nums, int k) {
                vector<int> tmp;
                unordered_set<int> hashSet;
                for (int num : nums) {
                        if (hashSet.find(num) == hashSet.end()) {
                                hashSet.insert(num);
                                tmp.push_back(num);
                        }
                }
                nums = tmp;
                sort(nums.begin(), nums.end());
                long long total = 0;
                for (int i = 0; i < nums.size(); i++) {
                        int emptySpot = nums[i] - (i+1);
                        if (emptySpot >= k) {
                                long long size = i + k;
                                long long totalSum = size * (size+1) / 2;
                                return totalSum - total;
                        }
                        total += nums[i];
                }
                long long size = nums.size() + k;
                long long totalSum = size * (size+1) / 2;
                return totalSum - total;
			}
	};

File: LeetCode root folder/Append K Integers With Minimal Sum.py
# greedy solution:55
# ms
# Beats
# 60.78%
class Solution:
    def minimalKSum(self, nums: List[int], k: int) -> int:
        nums = list(set(nums))
        nums.sort()
        total = 0
        for i in range(len(nums)):
            emptySpot = nums[i] - i - 1
            if emptySpot >= k:
                # print(nums[i], k)
                size = i + k  # size --> the size of the entire set of numbers 1,2,3...i+k, of which i of them already exist inside the input array (nums), and their sum is stored in 'total'.
                totalSum = (size) * (size + 1) // 2
                return totalSum - total
            total += nums[i]

        # since there are not enough empty spots for all k numbers in the range 1 to max(nums), we do the following
        size = len(nums) + k
        totalSum = (size) * (size + 1) // 2
        return totalSum - total




